datafusion/core/src/physical_optimizer/pruning.rs

```diff
@@ -233,25 +233,18 @@ impl PruningPredicate {
             .unwrap_or_default()
     }
 
-    /// Returns all need column indexes to evaluate this pruning predicate
-    pub(crate) fn need_input_columns_ids(&self) -> HashSet<usize> {
-        let mut set = HashSet::new();
-        self.required_columns.columns.iter().for_each(|x| {
-            match self.schema().column_with_name(x.0.name.as_str()) {
-                None => {}
-                Some(y) => {
-                    set.insert(y.0);
-                }
-            }
-        });
-        set
+    pub(crate) fn required_columns(&self) -> &RequiredStatColumns {
```

Comment:
This is the core change -- `need_input_columns_ids` returns indexes in terms of the overall table schema. If an individual parquet file does not have all the columns or has the columns in a different order, these indexes are not correct

Reply:
Thanks for explanation! 👍 
>  If an individual parquet file does not have all the columns or has the columns in a different order

I have a question about if `file_a (c1, c2), file_b(c1, c3)`, do df support create external table t(c1) on both file_a and file_b 🤔

Reply:
And file_a (c1, c2), file_b(c3) ,  support create external table t(c1)?
Do both file have to have the c1 meta in both parquet files meta ?

Reply:
i see both situation support in below test 😆